### ISSUE TITLE
make rumble:datum-intern-literal more optimizer friendly (draft)

### ISSUE THREAD

**Issue by gus-massa** [racket/racket#2073]
See the branch [gus-massa/18-5-RumbleIntern](https://github.com/gus-massa/racket/commits/18-5-RumbleIntern) in particular [gus-massa/1672af](https://github.com/gus-massa/racket/commit/1672af030df9404be730c6cf5ebb04c5afbae45c
).

The current implementation of `rumble:datum-intern-literal` is very optimizer unfriendly, but I don't have a good sugestion to fix it. This is a very early idea and I post it here in case someone can give any sugestions or feedback.

My idea is to add a check to ensure that the interned value is `equal?` to the initial value. This is ensured by the implementation but it's almost impossible for the optimizer to discover it. So the check acts as an optimization hint, without using a secret handshacke.

Is the `equal?` that I added `chez:equal?` or `rumble:equal?`?

But with the current version of ChezSheme+cptypes this is not enough. I expect to be able to fix this after cptypes is merged with three or four aditional small commits, and some tweaking of this change to `rumble:datum-intern-literal` to make cp0 and cptypes happy. (For example, move the `equal?`ity check, or split the implementation in two functions to make partial inlining easier. I'm really not sure.)

**Comment by mflatt** [comment]
I'm not clear on the intent of the change, but I can answer the easy question: that `equal?` is `rumble:equal?`, not `chez:equal?`.

Is your concern that `datum-intern-literal` makes it difficult for the optimizer to see that the same kind of value comes out as goes in? I can see how an `equal?` check would expose that property, but it seems like it could have a significant run-time cost.

**Comment by gus-massa** [comment]
Yes, I think that with the current version the optimizer can never guess the kind of value of the result. For example I expect some autogenerated code like

    (define temp (datum-intern-literal ""))
    (lambda (x) (equal? x temp))

and it would be nice in the future that the optimizer(s) in chez scheme knows that `temp` is actually a empty string. (This information is still difficult to use.)

It can also be helpful in functions that accept strings and bytes, and after inlining some of the internal checks can be removed in the optimization pass.

---

I don't think the additional check is too slow. I expect that the `weak-hash-ref-key` at the beginning of `datum-intern-literal` is as slow as the `equal?`. (There is an `eqv?` at the beginning of the implementation of `equal?` that should catch the values that are ignored by `datum-intern-literal`, and in strings and bytes `weak-hash-ref-key` must looks at all the characters to calculate the hash.)

Also, I don't expect to see too much uses of `datum-intern-literal` that are not lifted and run only once.
